```{r}
library(valr)
merged_igc <- merged_df %>%
    rename(
        `chrom` = "#reference_name.liftover",
        start = reference_start.liftover,
        end = reference_end.liftover,
    )
merged_igc$pvalue <- get_pvalues_from_igc(merged_igc)

all_sd <- rbind(SEDEF_V1.1, SEDEF_V1.1_LOWID, fill = TRUE) %>%
    rename(
        `chrom` = "chr",
    )
flat_all_sd <- as.data.table(reduce(toGRanges(all_sd)))
sum(flat_all_sd$end - flat_all_sd$start) / 1e6

hap_cov <- fread("data/haplotype_coverage_bed_graph.bed")
colnames(hap_cov) <- c("chrom", "start", "end", "hap_coverage")
# load TMRCA
tmrca <- fread("data/TMRCA/TMRCA.tbl")
# load gc
gc_df <- fread("data/gc_content/gc_content.bed")
```


# make 10 SD windows 
```{r}
merge_into_windows <- function(df, n_win = 10) {
    df[, bases := end - start]
    df[,
        sample_window := rep(1:.N,
            each = n_win,
            length.out = .N
        ),
        by = list(chrom, region, hap)
    ]

    grouped_windows <- df[,
        .(
            start = min(start),
            end = max(end),
            per_div = mean(per_div),
            bases = sum(bases),
            num_snv = sum(num_snv),
            n_windows = .N
        ),
        by = list(chrom, region, hap, sample_window)
    ] %>%
        filter(
            end - start > 0.9 * bases & end - start < 1.1 * bases
        ) %>%
        mutate(
            SNVsPer10k = 10e3 * num_snv / bases
        )
    grouped_windows
}
# add chrom name
snv_windows[, chrom := `#chr`]
# merge into 10 windows
windows_10 <- merge_into_windows(snv_windows, n_win = 10)
merged_windows <- windows_10[,
    .(
        hap_coverage = .N,
        num_snv = sum(num_snv),
        bases = sum(bases)
    ),
    by = list(chrom, start, end, region)
]
merged_windows[, SNVs_per_10 := num_snv / bases * 10e3]
# add gc info
columns <- colnames(merged_windows)
merged_windows <- merged_windows %>%
    bed_intersect(., gc_df, suffix = c("", ".gc")) %>%
    group_by_at(vars(one_of(columns))) %>%
    summarise(
        GC = mean(GC.gc),
        AT = 1 - GC
    ) %>%
    data.table()
merged_sd_windows <- merged_windows[region == "SD"]
# summary info
merged_windows %>%
    group_by(region) %>%
    summarise(
        mean(SNVs_per_10), n()
    )
```

# add igc or other info to SDs
```{r}
add_sd_info_to_df <-
    function(df,
             add_hap = FALSE,
             add_igc = FALSE,
             add_tmrca = TRUE,
             min_pvalue = 0.05) {
        columns <- c(
            colnames(df),
            "hap_coverage", "n_unique_igc",
            "AT", "GC", "TMRCA"
        )
        if (add_hap) {
            df <- df %>%
                bed_intersect(., hap_cov, suffix = c("", ".hap")) %>%
                group_by_at(vars(one_of(columns))) %>%
                summarise(hap_coverage = mean(hap_coverage.hap)) %>%
                as.data.table()
        }
        print(dim(df)) / 1e6
        if (add_tmrca) {
            df <- df %>%
                bed_intersect(., tmrca, suffix = c("", ".tmrca")) %>%
                group_by_at(vars(one_of(columns))) %>%
                summarise(
                    TMRCA = mean(TMRCA.tmrca),
                ) %>%
                data.table()
        }
        print(dim(df)) / 1e6
        print(colnames(df))
        if (add_igc) {
            counts <- as.data.table(
                findOverlaps(
                    toGRanges(df[, c("chrom", "start", "end")]),
                    toGRanges(merged_igc[pvalue <= min_pvalue, c("chrom", "start", "end")])
                )
            ) %>%
                group_by(queryHits) %>%
                summarise(N = n())
            df$n_unique_igc <- 0
            df[counts$queryHits]$n_unique_igc <- counts$N
        }
        print(dim(df))
        print(colnames(df))
        df <- df %>%
            bed_intersect(., all_sd, suffix = c("", ".SD")) %>%
            group_by_at(vars(one_of(columns))) %>%
            summarise(
                aln_len = mean(aln_len.SD),
                max_aln_len = max(max_len.SD),
                fracMatch = mean(fracMatch.SD),
                maxFracMatch = max(fracMatch.SD),
                sumFracMatch = sum(fracMatch.SD),
                SD_count = n()
            ) %>%
            data.table()

        df$n_unique_igc_per_hap <- df$n_unique_igc / df$hap_coverage
        df$has_igc <- factor(df$n_unique_igc_per_hap > 0)
        df$sd_divergence <-
            1 - df$maxFracMatch
        df
    }

merged_tmrca <- add_sd_info_to_df(
    merged_sd_windows,
    add_igc = TRUE,
    add_tmrca = TRUE,
    min_pvalue = 0.1
)
merged_sd_tmrca <- merged_tmrca[region == "SD"]

merged_windows_info <- add_sd_info_to_df(
    merged_sd_windows,
    add_igc = TRUE,
    add_tmrca = FALSE,
    min_pvalue = 0.1
)
merged_sd_windows_info <- merged_windows_info[region == "SD"]
```

```{r}
# note: linear mixed effects model
# note: nural net, might be predictive.
model <- lm(
    SNVs_per_10 ~
    (
        scale(log(SD_count)) +
            scale(sd_divergence) +
            scale(n_unique_igc) +
            scale(TMRCA) +
            scale(GC)
    )^2,
    data = merged_sd_tmrca
)
summary(model)
```



```{r}
merged_sd_windows_info %>%
    # filter(pvalue <= 0.05) %>%
    ggplot(aes(x = 100 * maxFracMatch)) +
    geom_histogram(binwidth = 0.25) +
    scale_y_continuous("# of IGC events (p < 0.05)", labels = comma) +
    scale_x_continuous(
        labels = comma,
        "Maximum percent identity of overlapping SD",
    ) +
    coord_cartesian(xlim = c(90, NA)) +
    theme_minimal_grid()
ggsave("tmp.pdf")
```

# SNVs per 10 vs SD identity 
```{r}
merged_sd_windows_info %>%
    # filter(pvalue <= 0.05) %>%
    ggplot(aes(x = 100 * maxFracMatch, y = SNVs_per_10)) +
    geom_hex() +
    # stat_cor() +
    geom_smooth(method = "lm", se = FALSE) +
    scale_y_continuous(
        "# SNVs per 10 kbp",
        labels = comma,
        trans = "log10"
    ) +
    scale_x_continuous(
        labels = comma,
        "Highest identity overlapping SD",
    ) +
    scale_fill_distiller("", palette = "Spectral", trans = "log10") +
    # coord_cartesian(xlim = c(90, NA)) +
    theme_minimal_grid()
ggsave("tmp.pdf")
ggsave("figures/SNVsPer10kbp_vs_SDidentity.pdf")
```
```{r}
model <- lm(
    n_unique_igc / hap_coverage ~
    (log(SD_count) +
        log(1 - maxFracMatch + 1 / max_aln_len) +
        max_aln_len +
        SNVs_per_10
    )^2,
    data = merged_sd_windows_info[n_unique_igc > -0]
)
summary(model)

merged_sd_windows_info %>%
    filter(n_unique_igc > 0) %>%
    filter(sd_divergence > 0) %>%
    ggplot(aes(x = 100 * (sd_divergence), y = n_unique_igc / hap_coverage)) +
    geom_hex(bins = 40) +
    stat_cor() +
    geom_smooth(method = "lm", se = FALSE) +
    scale_fill_distiller("", palette = "Spectral", trans = "log10") +
    scale_x_continuous("% divergence between SD paralogs", trans = "log10", label = comma) +
    scale_y_continuous("# unique IGC events per 1 kbp window per haplotype",
        # trans="log10",
        label = comma
    ) +
    annotation_logticks(side = "b") +
    theme_minimal_grid()
ggsave("tmp.pdf")
```




# merge igc and SD data
```{r}
columns <- c(colnames(merged_igc), "hap_coverage")

merged_igc_with_sd <- merged_igc %>%
    bed_intersect(., hap_cov, suffix = c("", ".hap")) %>%
    group_by_at(vars(one_of(columns))) %>%
    summarise(hap_coverage = mean(hap_coverage.hap)) %>%
    ungroup() %>%
    bed_intersect(., all_sd, suffix = c("", ".SD")) %>%
    group_by_at(vars(one_of(columns))) %>%
    summarise(
        aln_len = mean(aln_len.SD),
        max_aln_len = max(max_len.SD),
        fracMatch = mean(fracMatch.SD),
        maxFracMatch = max(fracMatch.SD),
    ) %>%
    data.table()
dim(merged_igc_with_sd)
```



```{r}
merged_windows %>%
    group_by(region) %>%
    summarise(mean(SNVs_per_10), n())
z <- merged_windows %>%
    filter(region %in% c("Unique", "SD")) %>%
    # filter(Superpopulation == "AFR") %>%
    mutate(
        GC_bin = cut(GC, breaks = seq(0, 0.60, 0.05))
    ) %>%
    filter(!is.na(GC_bin)) %>%
    group_by(region, GC_bin) %>%
    filter(n() > 2e3) %>%
    slice(sample(min(10e3, n()))) %>%
    ggplot(aes(x = GC_bin, y = SNVs_per_10, fill = region)) +
    geom_violin() +
    geom_label(
        data = . %>%
            group_by(GC_bin, region) %>%
            summarise(
                m = mean(SNVs_per_10)
            ),
        aes(y = m + 1 * (region == "SD"), label = round(m, 2)),
        color = "white"
    ) +
    # geom_hex() +
    # geom_smooth(method = "lm", se = FALSE) +
    # stat_cor() +
    # scale_fill_distiller("", palette = "Spectral", trans = "log10") +
    # facet_col(~region) +
    scale_y_continuous(
        "# of SNVs per 10 kbp\n(10,000 windows in each violin)",
        trans = "log10", label = comma
    ) +
    annotation_logticks(side = "l") +
    scale_fill_manual(values = c(SD = NEWCOLOR, Unique = OLDCOLOR)) +
    xlab("GC fraction") +
    theme_minimal_grid()
ggsave("tmp.pdf")
ggsave("figures/SNVs_per_10kbp_by_GC_fraction.pdf")
```
```{r}
library(karyoploteR)
pdf("tmp.pdf")
chrs <- NOYM
kp <- plotKaryotype(
    plot.type = 2,
    cytobands = CYTOFULL, genome = GENOME, chromosomes = chrs
)
ymax <- 2e4
kpPlotDensity(kp,
    data = toGRanges(gc_df[GC >= 0.6]),
    data.panel = 1, window.size = 100e3, col = transparent(NEWCOLOR, 0.25)
)
dev.off()
```